client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoResultSet.java

```diff
@@ -156,13 +158,13 @@ private static <T> Iterator<T> flatten(Iterator<Iterable<T>> iterator, long maxR
         // Semaphore to indicate that some data is ready.
         // Each permit represents a row of data (or that the underlying iterator is exhausted).
         private final Semaphore semaphore = new Semaphore(0);
-        private final CompletableFuture<Void> future;
+        private final Future<?> future;
 
         public AsyncIterator(Iterator<T> dataIterator, StatementClient client)
         {
             requireNonNull(dataIterator, "dataIterator is null");
             this.client = client;
-            this.future = CompletableFuture.runAsync(() -> {
+            this.future = executorService.submit(() -> {
```

Comment:
> This patch uses FutureTask, which
is created by ThreadPoolExecutor, instead of CompletableFuture to
make sure InterruptedException can be raised as expected.

I think relying on interrupts is inherently unreliable.
Should we also have explicit atomic/violatile boolean closed flag to ensure iteration is stopped even if some piece of code (our or lib) mishandles interruption?

Reply:
Thanks for the reply, @findepi . 

Using `Thread.interrupt` is a good choice, I think, because it can stop the work immediately when thread is waiting or sleeping. When thread is running, the InterruptedException will not be raised (I was wrong in the pr description), so we need to check the thread's `interrupted` status (it can act as the flag as you've mentioned) to make sure we can exit the iteration. I did not do the check here, so I will add it to make things right.

Reply:
Flag added, but still needs some improvement.

There are 2 test cases:
- One for queue is full and the thread will be blocked on `rowQueue.put`. Interruptions are useful to terminate the waiting. In this case, flag check is not working.
- One to simulate underlying implemetation mishandles interruptions. In this case, flag check can terminate the iteration.

It cannot solve the problem when situations are combined. For example, an interruption occurred (when cancel being invoked) during `dataIterator.hasNext()` or `dataIterator.next()` but has been mis-handled, and then the thread blocked on `rowQueue.put` for a full queue.

Reply:
Hi, @findepi. Things becomes complicated.

When we distruct the underlying implementation of interruption handling, there are cases we are not able to abort the running. For example, If the inner implemetation swalloweds the InterruptedException and still wait for more result data, the thread will not be released unless more result data generated. The flag check has nothing to do with the problem. 
Further more, after result data being generated, thread may be blocked at `rowQueue.put()` for a full queue. In this case, the thread will never be released. Addition code change, like using continuous `rowQueue.offer()`, is needed to avoid the long wait. 
Another problem is that when thread been terminated by a failure flag check, we cannot distinguish whether it is aborted or correctly finished. Some code should be added to make the distinction.

Here the issues is that threads may leak due to not be interrupted. The key problem is that `Thread.interrupt()` is never invoked during a `Future` cancellation. It is not a problem for interruption mishandle (at least for now) since the actual inner iteration implementation does do the right thing (maybe we can add some guard test cases to garantee). I think we can do a quick fix for this issue and to discuss about a new and better implemenation of TrinoResultSet in another thread.

For the reason above, I decide to revert the changes of add flag check and make the code change easy to understand.

Reply:
> Further more, after result data being generated, thread may be blocked at `rowQueue.put()` for a full queue. In this case, the thread will never be released. Addition code change, like using continuous `rowQueue.offer()`, is needed to avoid the long wait.

Why not drain (clear) the `rowQueue` upon cancellation?

Reply:
updated.